### PR TITLE
feat: rabbit errors reported to sentry

### PR DESCRIPTION
**Pull request by ReubenRobertsHEE**
TIS21-4925

### CODE REVIEW

**Comment by ReubenRobertsHEE** [comment]
Rabbit_sentry_issues

File: src/main/java/uk/nhs/hee/trainee/details/service/RabbitPublishService.java [hunk 1]
```diff
@@ -72,6 +118,12 @@ public void publishCojSignedEvent(ProgrammeMembership programmeMembership) {
       event = new CojSignedEvent(firstPmId, conditionsOfJoining);
     }
 
-    rabbitTemplate.convertAndSend(rabbitExchange, routingKey, event);
+    outstandingConfirms.putIfAbsent(programmeMembership.getTisId(), programmeMembership);
+    try {
+      rabbitTemplate.convertAndSend(rabbitExchange, routingKey, event, correlationData);
+    } catch (AmqpException e) {
```
Comment (ReubenRobertsHEE) [inline src/main/java/uk/nhs/hee/trainee/details/service/RabbitPublishService.java:124]: If Rabbit dies, catching the exception here means the front-end is not left hanging

File: src/main/java/uk/nhs/hee/trainee/details/service/RabbitPublishService.java [hunk 2]
```diff
@@ -51,6 +61,41 @@ public class RabbitPublishService {
...
+   *
+   * @param isAck           was the message acknowledged (success) or not (failure).
+   * @param correlationData message correlation details.
+   */
+  void handleRabbitAcknowledgement(boolean isAck, CorrelationData correlationData) {
+    final String id = correlationData.getId();
+    outstandingConfirms.remove(id);
+    if (!isAck) {
+      log.info("Rabbit message for programme membership id '{}' got nack-ed", id);
+      AmqpException e = new AmqpException("Rabbit message for programme membership id '"
```
Comment (ReubenRobertsHEE) [inline src/main/java/uk/nhs/hee/trainee/details/service/RabbitPublishService.java:90]: A more specific error might be useful e.g. [AmqpNackReceivedException](https://docs.spring.io/spring-amqp/api/org/springframework/amqp/rabbit/core/AmqpNackReceivedException.html) - but we don't have the message to-hand at this point.

File: src/main/java/uk/nhs/hee/trainee/details/service/RabbitPublishService.java [hunk 3]
```diff
@@ -51,6 +58,43 @@ public class RabbitPublishService {
     this.rabbitTemplate = rabbitTemplate;
   }
 
+  /**
+   * Set up the Rabbit confirmation callback handler.
+   */
+  @PostConstruct
+  public void postConstruct() {
+    rabbitTemplate.setConfirmCallback((correlation, ack, reason) -> {
+      if (correlation == null) {
```
Comment (ReubenRobertsHEE) [inline src/main/java/uk/nhs/hee/trainee/details/service/RabbitPublishService.java:67]: I've tried a few things, but I don't know how to test this effectively.

File: src/main/resources/application.yml [hunk 4]
```diff
@@ -19,13 +19,15 @@ spring:
...
+    publisher-confirm-type: correlated
+    max-message-age-unconfirmed: 10000
 
 logging:
   level:
     root: INFO
 
 sentry:
-  dsn: ${SENTRY_DSN:}
+  dsn: http://be5d80005ca84e62867f572b1f03c5da@localhost:9000/11
```
Comment (Judge40) [inline src/main/resources/application.yml]: Need to revert the change for local testing

File: src/test/java/uk/nhs/hee/trainee/details/service/RabbitPublishServiceTest.java [hunk 5]
```diff
@@ -121,10 +129,75 @@ void shouldPublishCojSignedEventWithUuid() {
...
+    Instant signedAt = Instant.now();
+    ConditionsOfJoining conditionsOfJoining
+        = new ConditionsOfJoining(signedAt, GoldGuideVersion.GG9);
+    programmeMembership.setConditionsOfJoining(conditionsOfJoining);
+
+    try {
+      rabbitPublishService.publishCojSignedEvent(programmeMembership);
+    } catch (Exception e) {
+      fail("Unexpected exception when Rabbit not available");
+    }
```
Comment (Judge40) [inline src/test/java/uk/nhs/hee/trainee/details/service/RabbitPublishServiceTest.java:201]: A cleaner approach would be to use `assertDoesNotThrow(() -> rabbitPublishService.publishCojSignedEvent(programmeMembership);, "Unexpected exception when Rabbit not available");`

**Review by Judge40 — CHANGES_REQUESTED** [review]

File: src/main/resources/application.yml (hunk 4 above)
Comment (ReubenRobertsHEE) [reply src/main/resources/application.yml]: Yes, sorry - at least its not a password or similar 😖

**Review by Judge40 — APPROVED** [review]
